### PR TITLE
services/horizon/internal/expingest: Add missing invariant assertions in the beginning of resume and init

### PR DESCRIPTION
**Pull request by tamirms**
<details>
  <summary>PR Checklist</summary>
  
### PR Structure

* [x] This PR has reasonably narrow scope (if not, break it down into smaller PRs).
* [x] This PR avoids mixing refactoring changes with feature changes (split into two PRs
  otherwise).
* [x] This PR's title starts with name of package that is most changed in the PR, ex.
  `services/friendbot`, or `all` or `doc` if the changes are broad or impact many
  packages.

### Thoroughness

* [x] This PR adds tests for the most critical parts of the new functionality or fixes.
* [ ] I've updated any docs ([developer docs](https://www.stellar.org/developers/reference/), `.md`
  files, etc... affected by this change). Take a look in the `docs` folder for a given service,
  like [this one](https://github.com/stellar/go/tree/master/services/horizon/internal/docs).

### Release planning

* [ ] I've updated the relevant CHANGELOG ([here](services/horizon/CHANGELOG.md) for Horizon) if
  needed with deprecations, added features, breaking changes, and DB schema changes.
* [ ] I've decided if this PR requires a new major/minor version according to
  [semver](https://semver.org/), or if it's mainly a patch change. The PR is targeted at the next
  release branch if it's not a patch change.
</details>

### What

Fixes https://github.com/stellar/go/issues/2204

* Add missing invariant assertions in the beginning of resume and init

### Why

https://github.com/stellar/go/issues/2204 provides some details on scenarios where we can have multiple versions of ingestion running at the same time during a rolling deployment.

If we have checks against the ingestion version found in the db during the init and resume state we can detect these cases and shutdown.

### Known limitations

[N/A]

### CODE REVIEW

File: services/horizon/internal/expingest/main.go [hunk 1]
```diff
@@ -623,6 +632,35 @@ func (s *System) resume() (state, error) {
...
+	}
+
+	if lastHistoryLedger != 0 && lastHistoryLedger != lastIngestedLedger {
+		return state{systemState: initState},
+			errors.Errorf(
+				"last history ledger %v does not match last ingested ledger %v",
+				lastHistoryLedger,
+				lastIngestedLedger,
+			)
+	}
```
Comment (bartekn) [inline services/horizon/internal/expingest/main.go]: Can you explain why is it necessary to check?

Comment (tamirms) [reply services/horizon/internal/expingest/main.go]: the init state ensures that either `lastHistoryLedger == 0` or `lastHistoryLedger == lastIngestedLedger` before the system can transition to `resume`. So this check makes that assumption explicit. If the condition does not hold it means something is wrong.

**Comment by tamirms** [comment]
> We probably need to lock access to order book graph if version mismatch is detected.

if the version does not match the ingestion system will stop running the resume state and it will go back to init. if the version is less than the current version it will rebuild state otherwise it will shut down. So, in both cases ingestion will stop adding offers to the orderbook graph until the db version is consistent with the current version

**Comment by bartekn** [comment]
Sorry, my message wasn't clear. I was thinking about API access to `/order_book` and `/paths`.

**Comment by tamirms** [comment]
> Sorry, my message wasn't clear. I was thinking about API access to /order_book and /paths.

I think the ingestion status check in the StateMiddleware will ensure that requests to the orderbook will fail: 

https://github.com/stellar/go/blob/release-horizon-v0.25.0/services/horizon/internal/middleware.go#L274-L278

**Comment by tamirms** [comment]
closed in favor of https://github.com/stellar/go/pull/2220

**Pull request closed without merging**